core/src/main/java/io/undertow/server/handlers/accesslog/DefaultAccessLogReceiver.java

```diff
@@ -153,75 +154,60 @@ public void logMessage(final String message) {
      */
     @Override
     public void run() {
+        //check if we can transition to 2. If so, perform tasks in small chunks and check this.closed.
+        //move into 3 if(this.closed) and terminate run()
         if (!stateUpdater.compareAndSet(this, 1, 2)) {
             return;
         }
-        if (forceLogRotation) {
-            doRotate();
+        //NOTE: once we are here, run() control state transition, unless it is too slow
+        //and close takes over after grace period.
+
+        if (forceLogRotation || System.currentTimeMillis() > changeOverPoint) {
+            performFileRotation();
         } else if (initialRun && Files.exists(defaultLogFile)) {
-            //if there is an existing log file check if it should be rotated
-            long lm = 0;
-            try {
-                lm = Files.getLastModifiedTime(defaultLogFile).toMillis();
-            } catch (IOException e) {
-                UndertowLogger.ROOT_LOGGER.errorRotatingAccessLog(e);
-            }
-            Calendar c = Calendar.getInstance();
-            c.setTimeInMillis(changeOverPoint);
-            c.add(Calendar.DATE, -1);
-            if (lm <= c.getTimeInMillis()) {
-                doRotate();
-            }
+            checkAndRotateOnInitialRun();
         }
-        initialRun = false;
-        List<String> messages = new ArrayList<>();
-        String msg;
-        //only grab at most 1000 messages at a time
-        for (int i = 0; i < 1000; ++i) {
-            msg = pendingMessages.poll();
-            if (msg == null) {
-                break;
+        if(closed) {
+            if (!stateUpdater.compareAndSet(this, 2, 3)) {
+                UndertowLogger.ROOT_LOGGER.accessLogWorkerFailureOnTransition();
             }
-            messages.add(msg);
+            return;
         }
+
+        //only grab at most 1000 messages at a time
         try {
-            if (!messages.isEmpty()) {
-                writeMessage(messages);
-            }
-        } finally {
-            // change this state to final state
-            stateUpdater.set(this, 3);
-            //check to see if there is still more messages
-            //if so then run this again
-            if (!pendingMessages.isEmpty() || forceLogRotation) {
-                if (stateUpdater.compareAndSet(this, 3, 1)) {
-                    logWriteExecutor.execute(this);
+            for (int i = 0; i < 1000 && !this.closed && !pendingMessages.isEmpty(); ++i) {
+                final String msg = pendingMessages.poll();
```

Comment:
It might be better to use peek()+remove() here.

Reply:
@baranowb because this class has concurrent access, poll is better than peek() + remove(). Between the moment we invoke peek() and the moment we invoke remove, we risk having another thread executing flushAndTerminate. This is why we need to run poll() here.

Reply:
There is trade off. It is possible that XNIO thread will be terminated between poll() and write(), leaking one message.
If its peek()+remove(), it might double entry, but not loose it. Its a trade off.

Reply:
You're right about it. My question is now is hot to prevent both scenarios. Is there some way we can synchronize just a little bit to prevent a duplicate entry in the log? I think we must prevent an output that will be considered buggy by users if someone finds a duplicate entry after shutting down the server.

Reply:
Not unless we control threat termination, which we dont.